packages/core/entity/FileDetail/RenderZarrThumbnailURL.ts

```diff
@@ -38,89 +38,131 @@ function transformAxes(originalArray: AxisData[]): TransformedAxes[] {
-        !Array.isArray(group.attrs.multiscales) ||
-        group.attrs.multiscales.length === 0
-    ) {
-        throw new Error("Invalid multiscales attribute structure");
-    }
-
-    // Access the image metadata and find lowest resolution.
-    const { multiscales } = group.attrs;
-    const datasets = multiscales[0].datasets;
-    const lowestResolutionDataset = datasets[datasets.length - 1];
-    const lowestResolutionLocation = root.resolve(lowestResolutionDataset.path);
-    const lowestResolution = await zarr.open(lowestResolutionLocation, { kind: "array" });
-
-    // Determine a slice of the image that has the best chance at being a good thumbnail.
-    // X and Y will have full range, while Z will be the middle integer.
-    // Non-spatial dims will default to None.
-    const axes = transformAxes(multiscales[0].axes);
-    const zIndex = axes.findIndex((item) => item.name === "z");
-    if (zIndex !== -1) {
-        const zSliceIndex = Math.ceil(lowestResolution.shape[zIndex] / 2);
-        axes[zIndex].value = zSliceIndex;
+async function retryWithTimeout<T>(fn: () => Promise<T>, retries = 3, timeout = 5000): Promise<T> {
+    let attempt = 0;
+    while (attempt < retries) {
+        try {
+            return await withTimeout(fn(), timeout);
+        } catch (error) {
+            attempt++;
+            console.warn(`Attempt ${attempt} failed. Retrying...`, error);
+            if (attempt >= retries) {
+                throw new Error(`Operation failed after ${retries} attempts: ${error}`);
+            }
+        }
     }
+    throw new Error("Unexpected error in retry logic");
+}
 
-    // Create a view (get data) of the determined lowest resolution using the
-    // choices made for each axes.
-    const lowestResolutionView = await zarr.get(
-        lowestResolution,
-        axes.map((item) => item.value)
+/**
+ * Helper function to enforce a timeout on async operations.
+ */
+async function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
+    const timeout = new Promise<never>((_, reject) =>
+        setTimeout(() => reject(new Error(`Operation timed out after ${ms}ms`)), ms)
     );
-    const u16data = lowestResolutionView.data as Uint16Array;
-
-    // Normalize Data to improve image visability.
-    const min = Math.min(...u16data);
-    const max = Math.max(...u16data);
-    const normalizedData = new Uint8Array(u16data.length);
-    for (let i = 0; i < u16data.length; i++) {
-        normalizedData[i] = Math.round((255 * (u16data[i] - min)) / (max - min));
-    }
+    return Promise.race([promise, timeout]);
+}
+
+/**
+ * Main function to attempt to render a usable thumbnail using the lowest
+ * resolution present in a zarr image's metadata.
+ */
+export async function renderZarrThumbnailURL(zarrUrl: string): Promise<string | undefined> {
+    try {
+        return await retryWithTimeout(
+            async () => {
+                console.log("Starting Zarr thumbnail rendering for:", zarrUrl);
```

Comment:
Do we still need the log here?

Reply:
resolved in [dcce3a5](https://github.com/AllenInstitute/biofile-finder/pull/268/commits/dcce3a53988fff6be5dbfdfd0dd0d19fcbda3357)